### ISSUE TITLE
Adding a wind gustiness parameterization

### ISSUE THREAD

**Issue by rljacob** [ESMCI/cime#1105]
The merge of PR #1065 proposed to introduce a gustiness parameterization.   This issue discusses how its currently done and how best to do it.

In module **seq_flux_mct**:  
In seq_flux_init_mct:  allocate prec_gust and set it to 0.

In seq_flux_atmocnexch_mct:  define gust_fac, set it to HUGE(1.0) and then overwrite with seq_infodata_GetData.   pass prec_gust and gust_fac to shr_flux_atmocn;  

In seq_flux_atmocn_mct:  set prec_gust to a2x%rAttr(index_a2x_Faxa_rainc,n);  define gust_fac, set it to HUGE(1.0) and then overwrite with seq_infodata_GetData (from drv_in).   pass prec_gust and gust_fac to shr_flux_atmocn

In module **seq_infodata_mod**
Add gust_fac to seq_infodata_type

In  seq_infodata_Init, declare gust_fac and add to /seq_infodata_inparm/ namelist block.   Set initial value to huge(1.0_SHR_KIND_R8), read value from seq_infodata_inparm namelist; copy value to infodata%gust_fac, 

In seq_infodata_GetData_explicit:    If present, gust_fac = infodata%gust_fac
In seq_infodata_PutData_explicit:  if present,    infodata%gust_fac  = gust_fac
In seq_infodata_bcast:  broadcast infodata%gust_fac
In seq_infodata_print:  print out gust_fac value.

In module **shr_flux_mod**
In shr_flux_atmOcn:  Add gust_fac and prec_gust to arguments.  
Define new function:
ugust(gprec) = gust_fac*log(1._R8+57801.6_R8*gprec-3.55332096e7_R8*(gprec**2.0_R8))
Alter vmag calculation to include gustiness:
vmag   = max(umin, sqrt( (ubot(n)-us(n))**2 + (vbot(n)-vs(n))**2) +  ugust(min(prec_gust(n),6.94444e-4_R8)))

**Comment by mvertens** [comment]
The key issue that I am concerned about is that cesm never use the
gustiness parameterization - not that its values happen to be 0 by default.

**Comment by mvertens** [comment]
I feel that since this is a new parameterization that ACME has introduced into share science code that also happens to be a key to CESM coupling, there should be an if-block that ensures that CESM does not use it - even if a user in seq_flux_mct.F90 changes the gustiness values to be non-zero.
The way to do this is to pass cime_model to shr_flux_mct.F90 and have the following type of if-block

```
if (trim(cime_model) == 'acme') then
        vmag   = max(umin, sqrt( (ubot(n)-us(n))**2 + (vbot(n)-vs(n))**2) + ugust(min(prec_gust(n),6.94444e-4_R8)))
else
        !--- vmag+ugust (convective gustiness) Limit to a max precip 6 cm/day = 0.00069444 mm/s.
        !--- reverts to original formula if gust_fac=0 
        vmag   = max(umin, sqrt( (ubot(n)-us(n))**2 + (vbot(n)-vs(n))**2)
end if


```

**Comment by gold2718** [comment]
Since this is a new issue, I will reiterate my concern with the proposed solution and my suggestion:

Every other science option is bracketed by logic, I do not see why this one is different. If folks do not like having model names (e.g., cesm & acme), then come up with a namelist parameter which indicates the scientific reason for the change and use that.

For instance, following usage in physics packages, introduce an XML variable, `DO_GUSTINESS` and an associated namelist parameter, `do_gustiness`, which will be always off in CESM (set by default in config_component_cesm.xml) and could always be on for ACME (config_component_acme.xml).

This makes the science choices explicit, just like other parameterization choices in the models.

**Comment by mvertens** [comment]
I think this is better than what I have suggested and I am happy to go with
this implementation.

**Comment by rljacob** [comment]
There already is logic to turn this off;  if gust_fac is 0, its off.   gust_fac=0 will make the ugust function evaluate to 0.  So it functions as both a DO_GUSTNISS logical and a value in the parameterization.

**Comment by gold2718** [comment]
@rljacob, why are you insisting on changing how parameterizations have always been introduced? To take just one example, in the atmosphere physics package, the CLUBB SGS parameterization is bracketed by `if(do_clubb_sgs)`, not by magic numbers.
It appears to me that you are upending many years of best practice by hiding a parameterization in math rather than by explicit code logic. Why do you feel this is important in this case? What SE principle is involved?

**Comment by rljacob** [comment]
If you're choosing between two large blocks of code or calling a different subroutine, then yes something like if(do_gustiness) would be appropriate.   But this is just one term in one line so gust_fac=0 or not is the more elegant and better solution.  We can clarify with some inline documentation and the documentation for gust_fac.

**Comment by gold2718** [comment]
Elegance is in the eye of the beholder and is not a software-engineering principle.
While the code is a bit shorter in this case, logically and scientifically, it is very similar to the introduction of the flux_diurnal parameterization.
Why do you feel that a departure from the standard protocol is "better" in this case?

**Comment by mvertens** [comment]
First - Introducing new parameterizations - as this is - has always been done with introducing namelist input - as is the case with flux_diurnal. You can see this in seq_mct.F90.

Second, the default behavior of this parameterization was introduced in a way that was non backwards compatible and could easily have changed science since prec_gust was not initialization to zero -
` prec_gust (n) = a2x%rAttr(index_a2x_Faxa_rainc,n)
`
Even if this was set to zero now, you could see a case in the future where someone changed this back to non-zero and it would not be caught in testing if other changes were introduced in other models and we we could differences in science without even being aware of it.

Relying on hard-wiring prec_gust and gust_fact to zero in the code and assuming that this will always stay that way is not a robust way to handle a critical piece of science code in my opinion.



 - from examining the code carefully, this is actually not correct. It ugust is in fact zero - but prec_gust is NOT (see line 1397 in seq_flux_mct.F90)
prec_gust (n) = a2x%rAttr(index_a2x_Faxa_rainc,n)

**Comment by mvertens** [comment]
In looking at the code some more, I think the situation is much more complicated than simply setting an array to 0. I will outline my understanding below.

**ugust in shr_flux_atmOcn.F90 is a function - NOT an array**

There is a ugust(:) array in seq_flux_mct.F90 - but it has always been there and used as an input to the shr_flux_atmOcn_diurnal routine and was not used. Also-  cesm is not using shr_flux_atmOcn_diurnal in its current simulations.

ugust in shr_flux_atmOcn is the following function:

`ugust(gprec) = gust_fac*log(1._R8+57801.6_R8*gprec-3.55332096e7_R8*(gprec**2.0_R8))
`

`gust_fac `is being sent to `shr_flux_atmOcn` and this is being used to evaluate the function `ugust`.  

By default gust_fac is hug(1._r8) (this is the default value in shr_infodata_mod.F90).

Even if gprec is 0._r8, if you do not explicitly set `gust_fac` in the namelist to a value other than `huge(1._r8) `then this expression will evaluate to NaN if you compile it with debug on. (I tried this in a quick test).

I hopefully am getting this right - but is not behavior that we can have in cesm. I'm not sure how acme is handling this other than always having to have a dust_fac namelist option that is not huge(1._r8).

Please let me know if I am missing something.

**Comment by rljacob** [comment]
I agree setting the initial value to HUGE(1.0) can be removed.  That seems like an overly cautious way to make sure the namelist setting is taking effect.

**Comment by rljacob** [comment]
"flux_dirunal" is different because that logical controls if a completely different subroutine is called, one that has dozens of lines different from the non-diurnal version.  This is one term in one line of code.

**Comment by jgfouca** [comment]
@rljacob @mvertens @gold2718 , looking at this conversation, it's not clear to me what the resolution is.

**Comment by rljacob** [comment]
Go ahead and merge it as-is.  @mvertens will then do a PR with an implementation for CESM.

**Comment by mvertens** [comment]
It would help if just the following two minor changes were made to this PR.
This  would ensure that gust_fac is always being initialized as zero.

I think that for the purposes of this PR all that needs to be done is:

1) Change line 486 of seq_infodata_mod.F90

       gust_fac  = huge(1.0_SHR_KIND_R8)   <==== current
       gust_fac  = 0._r8 <==== new

at line 486 of seq_infodata_mod.F90.

2) Change line 910 of seq_flux_mct.F90:

    real(r8)    :: gust_fac = huge(1.0_r8) !wind gust factor    <====
current
    real(r8)    :: gust_fac     <==== new

since you are obtaining gust_fac from seq_infodata in line 910 of
seq_flux_mct.F90.

    call seq_infodata_GetData(infodata, &
         read_restart=read_restart, &
         dead_comps=dead_comps,         &
         atm_nx=atm_nx, atm_ny=atm_ny,  &
         ocn_nx=ocn_nx, ocn_ny=ocn_ny,  &
         ocn_prognostic=ocn_prognostic, &
         flux_diurnal=flux_diurnal,     &
         gust_fac = gust_fac            )

Moving forwards, we will be getting input from CESM scientists as to how
they want to see this science change implemented to satisfy CESM
requirements. However, that can be put into a upcoming PR.

**Comment by rljacob** [comment]
Sure I'll make those changes.

**Comment by singhbalwinder** [comment]
Sorry, just saw this conversation. Implementation is not complete yet as currently I don't know a way to turn this on based on a particular compset. Originally, I wanted to turn this on (gust_fac>0) only for a new compset but at that time, I couldn't quite figure out how to do it. I think I talked with @rljacob at that time but I don't remember if we were able to find a solution. 

This was my first time to make changes in the coupler codes and it sounds like I might have broken some existing rules. I will try to answer based on how I understand it works.

1. `real(r8)    :: gust_fac = huge(1.0_r8)`: I generally do that so that if this variable is used before it is initialized the simulation breaks (to be extra cautious). In this case, it is initialized right after so we can remove this.  I also realize that it will be an automatic `save` variable (it is not required to be `save` and was not intentional) so I think it is a okay to just declare it without initializing. I followed `flux_diurnal` variable to add this variable.

2.  `In seq_flux_init_mct: allocate prec_gust and set it to 0.`: I followed `prec` variable to add this variable. `prec` was also allocated and set to zero just after that, so I just followed suit. This can also be changed if necessary.

Rest of the code is just to read gust_fac from the namelist and pass it along so that it can be used in shr_flux_atmOcn. I can re-organize the code if it doesn't fit the current way of adding code to the coupler.

After implementing this code, I ran tests to make sure the code stays BFB with the old code.
